casper/src/test/scala/coop/rchain/casper/util/comm/CasperPacketHandlerSpec.scala

```diff
@@ -233,46 +239,197 @@ class CasperPacketHandlerSpec extends WordSpec {
+          _                   = assert(casperO.isDefined)
+          blockO              <- blockStore.get(genesis.blockHash)
+          _                   = assert(blockO.isDefined)
+          _                   = assert(blockO.contains(genesis))
+          handlerInternal     <- refCasper.get
+          _                   = assert(handlerInternal.isInstanceOf[ApprovedBlockReceivedHandler[Task]])
+          // assert that we really serve last approved block
+          lastApprovedBlockO <- LastApprovedBlock[Task].get
+          _                  = assert(lastApprovedBlockO.isDefined)
+          approvedBlockReq   = ApprovedBlockRequest("test")
+          approvedBlockPacket = Packet(transport.ApprovedBlockRequest.id,
+                                       approvedBlockReq.toByteString)
+          approvedBlockRes <- casperPacketHandler.handle(local)(approvedBlockPacket)
+          _ = assert(
+            approvedBlockRes.map(p => ApprovedBlock.parseFrom(p.content.toByteArray)) == Some(
+              lastApprovedBlockO.get))
+        } yield ()
 
-      "make a transition to ApprovedBlockReceivedHandler once ApprovedBlock has beeen received" in {
-        pending
+        test.unsafeRunSync
       }
 
       "stop querying peers once ApprovedBlock has been received" in {
-        pending
+        implicit val ctx = TestScheduler()
+        val fixture      = setup()
+        import fixture._
+
+        transportLayer.setResponses(_ => p => Right(p))
+
+        implicit val ph = new PacketHandler[Task] {
+          override def handlePacket(peer: PeerNode, packet: Packet): Task[Option[Packet]] =
+            Task.now(None: Option[Packet])
+        }
+
+        implicit val lab = LastApprovedBlock.unsafe[Task]()
+        val interval     = 1.second
+
+        assert(transportLayer.requests.size == 0)
+        val test = for {
+          _ <- CommUtil.requestApprovedBlock[Task](interval).forkAndForget
+          _ = ctx.tick()
+          _ = assert(transportLayer.requests.size == 1)
+          _ = ctx.tick(interval)
+          _ = assert(transportLayer.requests.size == 2)
+          _ <- lab.set(ApprovedBlock())
+          _ = ctx.tick(interval)
+          _ = assert(transportLayer.requests.size == 2)
+          _ = ctx.tick(interval)
+          _ = assert(transportLayer.requests.size == 2)
+        } yield ()
+
+        test.unsafeRunSync
       }
     }
 
     "in ApprovedBlockReceivedHandler state" should {
+      import monix.execution.Scheduler.Implicits.global
+      val fixture = setup()
+      import fixture._
+
+      val runtimeDir     = BlockStoreTestFixture.dbDir
+      val activeRuntime  = Runtime.create(runtimeDir, 1024L * 1024)
+      val runtimeManager = RuntimeManager.fromRuntime(activeRuntime)
+
+      val genesis                = HashSetCasperTest.createGenesis(Map.empty)
+      val approvedBlockCandidate = ApprovedBlockCandidate(block = Some(genesis))
+      val approvedBlock: ApprovedBlock = ApprovedBlock(
```

Comment:
This looks pretty similar to the approved block created above. Maybe extract creation out into a convenience method.